### ISSUE TITLE
Freestyle Libre 3 support

### ISSUE THREAD

**Issue by Cybereu** [NightscoutFoundation/xDrip#1474]
Recently Freestyle Libre 3 was launched in Europe providing GCM, readings are now through Bluetooth instead of NFC.
Is there any move to support it?
Thanks

* https://abbott.mediaroom.com/2020-09-28-Abbotts-FreeStyle-R-Libre-3-System-Receives-CE-Mark-Features-Worlds-Smallest-Thinnest-Sensor-with-Best-in-Class-Performance-at-the-Same-Low-Cost-for-People-with-Diabetes

**Comment by tolot27** [comment]
Abbott just announced the Libre 3. It is not available yet. Hence, nobody can start implementing support or investigate whether it is possible or not.

**Comment by schrackschorsch** [comment]
libre 3 is now available in Germany. The libre-app for ios is ready. The app for android soon will be available.
[https://www.freestylelibre.de/libre/freestyle-libre-3-entdecken.html?utm_source=newsletter&utm_campaign=NPI|DE_FSL3_LAUNCH_PATIENTEN|Email&utm_medium=email&M_BT=1164167858090](url)

**Comment by gui-dos** [comment]
I noticed that the NFC manufacturer code changed from `07` (Texas Instruments) to `7A` (Abbott Diabetes Care USA), therefore the NFC custom commands have to be changed accordingly.

**Comment by marekmand** [comment]
Shouldn't be this ticket reopened since FL3 is available publicly in Germany?
I don't see any reason to postpone this anymore.

**Comment by renele** [comment]
any update on this or where can we track progress or is something blocking ?

**Comment by yoshimo** [comment]
The missing android app makes analysis harder

**Comment by tolot27** [comment]
It is unlikely that xDrip will support L3, natively. Likely, it will be implemented similar to L2 OOP external app.

**Comment by birdfly** [comment]
> https://apkdownload.com.de/apk/1525101160/freestyle-libre-3-de
> 
> Here is it - Not officially Release until now...

but the link is not available, better share apk directly if you have it

**Comment by Ficsy76** [comment]
https://play.google.com/store/apps/details?id=com.freestylelibre3.app.de

**Comment by hoppelEwa** [comment]
Good morning,
is there a Libre3 support in xDrip by now? I used the patched FL2 app and ordered the testpackage. Can I just use it the same way?

**Comment by yoshimo** [comment]
Research is still ongoing.
The official Freestyle 3 app on android is available but not officially released yet according to the support.
It is kind of unstable and wasn't patched yet.
You can't use the old FSL1+2 App for the new sensor.

**Comment by tkcgm** [comment]
https://play.google.com/store/apps/details?id=com.freestylelibre3.app.de&gl=DE

**Comment by PieterDeBruijn** [comment]
In The Netherlands the Libre 3 is now 2 weeks on the market, officially released.
https://play.google.com/store/apps/details?id=com.freestylelibre3.app.nl
Version of the app/apk is v3.3.0.

**Comment by dassystem** [comment]
There is a program to gather data from Libre 3 to Nightscout via Libre Link: https://github.com/timoschlueter/nightscout-librelink-up 

Although I prefere an offline solution maybe we could consider this workaround to grab the data from the official Freestyle Libre servers.

**Comment by BPavol** [comment]
I needed to use official FreeStyle Libre 3 app outside Germany and on unsupported device. So I was looking for solution how to install this app outside Google Play. I have managed to repack official app with modified source where I disable this check. I wrote tutorial with detailed instructions how to do it:
https://rentry.co/t9ee6

I did not try it with sensor because my phone does not have NFC(I did not know I need that). But tutorial and UI of app works as expected. So I will be glad if those knowledges help someone with implementing support into xDrip.

**Comment by gfornax** [comment]
> There is a program to gather data from Libre 3 to Nightscout via Libre Link: https://github.com/timoschlueter/nightscout-librelink-up

I can confirm this works to some extent after a few minor fixes to get it running on the same host.
However, current xdrip seems to struggle with the every minute readings of the libre3. NS shows it correctly, but xdrip as a follower will show several different readings at the same (!) time and leave gaps of a few minutes. This also messes up the delta calculation, throws random warnings and crashes other connected apps (here: garmin data fields). Anyway, still better than the original Libre App.

I second a native bluetooth integration is the way to go. Not sure about the current status, I'd be happy to participate if that helps.

**Comment by charliebravoAT** [comment]
Is there any progress on that since the ticket is closed again?

**Comment by tolot27** [comment]
As I mentioned earlier, there will be no direct support for Libre 3 from xDrip because the L3 has no public API to read BG data and Abbott will shutdown our repository if we implement reverse-engineered stuff.

Hence, no need to ask again. I am sorry to tell you this.

**Comment by charliebravoAT** [comment]
@tolot27 thanks for the really fast response!
Oh no, so there is (and will be) no way to send the values to my watch? Except via Nightscout, what will be a problem since I travel a lot also in remote areas with no internet :O
Oh man, too sad... I don't understand abbott...

**Comment by tolot27** [comment]
@charliebravoAT You can still use the Abbott app to get the data from the L3. This works without internet connection, as far as I know. At least with L2 no internet connection was required.
Another option would be the Supersapiens connect to Garmin devices: https://supersapiens.zendesk.com/hc/en-us/articles/1500008411202-Connect-Your-Garmin-Watch-to-Supersapiens

**Comment by Klaus3d3** [comment]
I'm pretty sure you could at least grab the notification from the freestyle app with tasker or so and inject these values into xdrip.

**Comment by charliebravoAT** [comment]
@tolot27 Yeah, but the Abbott app has no function to give it to my watch (or is there a way?)
The thing is (I know, xdrip developers can't do much about that), especially if I'm in the middle of nowhere (an example, a 10 day hike in the Andes), there is no way to use nightscout. And especially during such hikes, the value on my watch is really helpfull. Much more better than having to take out my phone and having a look on that every 30 mins with sweaty and/or dirty hands.
Garmin may be an option... but not with my watch (a super cheap, but practial MiBand, with more than 14 days batterylife :D )

So however, I'll adress Abbott (it might not help, but if enough ppl are asking, they may change sth, the other option would be not to take Abbott any more)...

OR
Let's hope somebody will develop a patched Libre 3 App :D :D 

@Klaus3d3 I think the L3 App only gives notifications when the value is to high or to low, not in range, right?
But I might be wrong.

**Comment by gfornax** [comment]
> I think the L3 App only gives notifications when the value is to high or to low, not in range, right?
> But I might be wrong.
Almost. I had a constant notification banner that says something like "unlock to see BG value", but it does not show your current reading, neither on the lock screen nor unlocked. You have to open the app, which is really annoying.
The high/low warning notification banners, however, will show the current value, even on the lock screen.

**Comment by tzachi-dar** [comment]
Can the high/low thresholds be configured to always create an alert?
For example, set the low alert trigger to 400, and then the low alert will always be triggered (I hope).

**Comment by charliebravoAT** [comment]
@tzachi-dar that may be an Option ^^

Set an alarm to get notifications
Send that notifications to Xdrip
Create "real" alarms in xdrip (otherwise you don't have real alarms)

**Comment by Klaus3d3** [comment]
> Can the high/low thresholds be configured to always create an alert? For example, set the low alert trigger to 400, and then the low alert will always be triggered (I hope).

Or maybe just eat a lot of ice cream. A very yummy workaround.

**Comment by Klaus3d3** [comment]
> > I think the L3 App only gives notifications when the value is to high or to low, not in range, right?
> > But I might be wrong.
> 
> Almost. I had a constant notification banner that says something like "unlock to see BG value", but it does not show your current reading, neither on the lock screen nor unlocked. You have to open the app, which is really annoying. The high/low warning notification banners, however, will show the current value, even on the lock screen.

And they call it a cgm? I think they missed the spot.

**Comment by gfornax** [comment]
> Can the high/low thresholds be configured to always create an alert?
> For example, set the low alert trigger to 400, and then the low alert will always be triggered (I hope).

I tried that for exactly that reason :) It works, but I could not determine what exactly triggers the existing notification to be refreshed/replaced with the current reading. It will certainly not happen automatically ever minute.

**Comment by Teute0815** [comment]
Wouldn't it be possible to add
LibreView and/or
LibreLinkUp
accounts as possible Data sources in xDrip in order to receive the values?

That was not feasible with L2, as that didn't read/send the values itself without scanning, but the L3 does that automatically every minute.

Even though it requires internet connection it would be a first/quick and easy to use "workaround" until there's a proper/easy (offline) solution

Atm there are a lot of users who can't/don't want to use the other available solutions - of which I found three so far:
*rooted phone - "XPosed/LSPosed Modul für Libre 3"
*LibreLinkUp "collector" apk
*Nightscout script/plugin - Follower

**Comment by tkcgm** [comment]
use websearch, search for "freethree" or "freethree package", 
root your device
use libre3 and  xdrip+ local on your device 
no librelinkup or so needed, sam e use like libre2

lets party

**Comment by Teute0815** [comment]
> use websearch, search for "freethree" or "freethree package", root your device use libre3 and xdrip+ local on your device no librelinkup or so needed, sam e use like libre2
> 
> lets party

I've mentioned the freethree in my post already^^
It's something which is just not doable for everybody. Data through LibreView/Up would be a first quick solution to keep going.

**Comment by gui-dos** [comment]
Let's mention @DecentWoodpecker67 explicitly: hopefully the author will open a public repo...  ;-)

**Comment by Joerg-Schoemer** [comment]
> As I mentioned earlier, there will be no direct support for Libre 3 from xDrip because the L3 has no public API to read BG data and Abbott will shutdown our repository if we implement reverse-engineered stuff.
> 
> Hence, no need to ask again. I am sorry to tell you this.



> > There is a program to gather data from Libre 3 to Nightscout via Libre Link: https://github.com/timoschlueter/nightscout-librelink-up
> 
> I can confirm this works to some extent after a few minor fixes to get it running on the same host. However, current xdrip seems to struggle with the every minute readings of the libre3. NS shows it correctly, but xdrip as a follower will show several different readings at the same (!) time and leave gaps of a few minutes. This also messes up the delta calculation, throws random warnings and crashes other connected apps (here: garmin data fields). Anyway, still better than the original Libre App.
> 
> I second a native bluetooth integration is the way to go. Not sure about the current status, I'd be happy to participate if that helps.

I've made some changes to Timo’s librelink up updater https://github.com/timoschlueter/nightscout-librelink-up
The interval is now 5 min, to meet the needs from nightguard.

**Comment by PieterDeBruijn** [comment]
Great!

We should definitely have Nightscout be compatible with 1 minute
measurements. Is there already a feature request for this?

Regards, Pieter

Op ma 9 mei 2022 17:55 schreef Jörg Schömer ***@***.***>:

**Comment by Teute0815** [comment]
@AlFontal 

Here you go. It's a german diabetes site, but that's the guy who made it:
[https://insulinclub.de/index.php?thread/33795-free-three-ein-xposed-lsposed-modul-f%C3%BCr-libre-3-aktueller-wert-am-sperrbildschir/](url)

**Comment by AlFontal** [comment]
Thanks @Teute0815, that works!

**Comment by twinko** [comment]
Hi everyone,

again a link to a german website but here someone createt somehing that doesnt need Root. Only requirement= internet connection:
https://insulinclub.de/index.php?thread/33987-llu-client/&postID=654168

> I have created an Android app that fetches the current glucose value every minute via LibreLinkUp (=from the LibreView servers) and displays it as a notification and thus also on the lockscreen. At the same time, the value is broadcast to xDrip+. In xDrip+, "Libre2 (patched app)" must be selected as the data source.


And to summarize htis thread. For now im forced to root (not possible for mw due to banking apps), use the llu client or have to wait for a patched libre 3 app. This is so bad from abbott :(

**Comment by jw243** [comment]
It seems to me that with the web follower source, the glucose values cannot be calibrated. Or am I missing something?
The llu-client on the other hand seems to work quite well, as long as you exempt it from battery optimisation. Here, the raw data can nicely be calibrated. What a game changer compared to the app from Abbott.

**Comment by twinko** [comment]
just for the people readingf this, look here, there is a solution now:
https://github.com/maheini/FreeStyle-Libre-3-patch

**Comment by emmatovar27** [comment]
> just for the people readingf this, look here, there is a solution now: https://github.com/maheini/FreeStyle-Libre-3-patch

Sadly this doesn't work with the sensors from the US! :( If there is any chance to replicate the same patch I will be happy to help! :)

**Comment by gringyb** [comment]
I have been using the ghost libre 3 app. It is completely offline and does not even ask for android internet permission. There is nothing to configure to get it working. I just scan my existing sensor and set xdrip to use patched app. I think it is based on the good juggluco work as it shows that name in the app menu. I download it from: https://mega.nz/file/7kJAzABT#3FGA60327a5w3E_VPoXCRXQjj1w_QG2lt-Dn_1fMcg8

**Comment by Nb-tst** [comment]
> 

did not know that such an app existed, is there an (un)official page for refrence?

**Comment by twinko** [comment]
Would be very interested in the source of ghost libre3 as well.

I personally **wouldn't recommend anybody to download and install the app** until we have more information about it. @gringyb is a freshly made account and i couldn't find anything online. Be carefull everyone

**Comment by charliebravoAT** [comment]
> just for the people readingf this, look here, there is a solution now: https://github.com/maheini/FreeStyle-Libre-3-patch


Since there is a way from the patched libre 3 App to juggluco, do you think there might be way to get it directly to xdrip?
So without juggluco (and not the ghost app).
So that xdrip gets the values the same way like juggluco:)
Woul be amazing 😍

**Comment by gringyb** [comment]
@charliebravoAT This is what the ghost app does. I have been using it no problem for more than 2 week. It is libre app patched to send direct to xDrip - Same like old libre2 patched app it has no internet permission and so cannot do anything not wanted, break privacy or mess with other android apps because of normal android security feature. It is made on top of juggluco patching method but not needing juggluco app or any special configuration to work. The only problem I have seen is some time needing to turn bluetooth off then on if signal lost. If people do not trust it then get an expert to look at it rather than assume bad things because we like to be private.

**Comment by emmatovar27** [comment]
Do you know if it works with FSL 3 sensors from the US?

**Comment by dassystem** [comment]
> @charliebravoAT This is what the ghost app does. I have been using it no problem for more than 2 week. It is libre app patched to send direct to xDrip - Same like old libre2 patched app it has no internet permission and so cannot do anything not wanted, break privacy or mess with other android apps because of normal android security feature. It is made on top of juggluco patching method but not needing juggluco app or any special configuration to work. The only problem I have seen is some time needing to turn bluetooth off then on if signal lost. If people do not trust it then get an expert to look at it rather than assume bad things because we like to be private.

Happy to hear that it is working for you,

Since we are building open source software, we want to see sources of the programs we should use. We need source code to check correct behavior and security, but also to be able to integrate it in our systems like xdrip or nightscout. And we need open source to be able to react to hardware related changes or a new encryption from the manufacturer. Also we need independence from single individuals building great software like juggluco or maybe ghost app. We need a reliable and qualified community.

I personally haven't seen sources of ghost app OR juggluco and this really concerns me. Even when they said, jugglucos libre patch can easily decompiled, I still prefer a clean repo with all code and instruction needed, so that I can build the software myself in case of need.

**Comment by twinko** [comment]
@dassystem in this specific case (patching libre app)  there is an obvious reason why there is no repo. Abbott would shut it down and have there lawyers send a letter.
Take in mind the situation with the patched libre2 app.

**Comment by gui-dos** [comment]
@dassystem:
@ivalkou’s https://github.com/ivalkou/LibreTools was the first repo to unveil the native methods and the keys to XOR-decrypt the Libre 2 FRAM and it is still online as well as the other repos detailing the reversed Bluetooth protocols and the temperature-based calibration algorithm...

**Comment by twinko** [comment]
did anybody tried the ghost app?

**Comment by malloy139** [comment]
I tried the jka modified Libre 3 app -> juggluco -> xdrip solution. I really don't like that another app needs to run all the time, besides the strange look and feel of juggluco. In xdrip you can see a new value every minute, they are also much better and usable than Libre 2. My problem now is the excessive smoothing of the raw values xdrip applies. This would lead to delayed alarms by xdrip, even with raise an alarm based on forecast. For this reason I use the alarms of the Libre app.
On the xdrip follower sadly you can't see the raw values.
Does anyone know if there is an option to sync raw values? There exists a pull request and issue. Also is it possible to turn off the smoothing of the values?

When the sensor ends next week I could try the ghost app.

**Comment by twinko** [comment]
@malloy139 
sorry i cant finde the source right now, but i read somewhere that the smoothing happens in the libre 3 sensor now, so there is no way to "de-activate" it. I probably read it here somewhere (german forum: https://insulinclub.de/index.php?thread/34255-libre3-2-juggluco/&pageNo=26)
 If xdrip is applying an additional smoothing, please send the link to the pull request to read more about that.

Regading the ghost app: would be interested to hear about your experience,

**Comment by malloy139** [comment]
@twinko juggluco is sending a raw value every minute and xdrip applies smoothing to it and gives you a new value only every 5 minutes. This is quite inconvenient when the BG changes quickly, see the attached image.
![58b62dc0-afa1-4f67-8a66-26a79740f371_1](https://user-images.githubusercontent.com/41596803/202567467-da19c561-de57-4234-bd4c-6ec18f37e4a7.jpg)


Syncing raw values, there is this issue: #1597  and these PR: #1686 and #1821. The last one says it implements this but it does not.

**Comment by twinko** [comment]
@malloy139 this is the wrong thread to address this problem. But the pr that needs to get merged to get this feature is the following one:
https://github.com/NightscoutFoundation/xDrip/pull/1686
Please comment there (like I did) perhaps this bumps it up a little :)

**Comment by harentius** [comment]
@twinko Looks like #1686 doesn't solve this problem. It syncronize followers, but doesn't solve issue with per 4.5 min xdrip plot, and, I assume, sending data to AndroidAPS via local broadcast.

**Comment by twinko** [comment]
@gringyb tried the ghost app. Doesn't launch, crashes immediately. Wiping my phone now and restore my backup. Afterwards i tried jkaltes (juggluco) patched app. This one is working like a charm. So i guess its not my phone that is the problem.

**Comment by master3395** [comment]
> @gringyb tried the ghost app. Doesn't launch, crashes immediately. Wiping my phone now and restore my backup. Afterwards i tried jkaltes (juggluco) patched app. This one is working like a charm. So i guess its not my phone that is the problem.

I got the same issue.

Both LL Patched and Juggluco works fine, but juggluco does not send all values to xDrip+ more than once each 5 min or so.

**Comment by j-kaltes** [comment]
xDrip itself transforms values coming in every minute to one value very 5 minutes. 
Take a new version of xDrip and turn on _Settings->Less Common settings->Advanced settings for Libre2->"Show Raw values in Graph"_ and you will see that every minute a new "raw" value arrives in xDrip.

**Comment by master3395** [comment]
> 

Jiggluco sends all the raw data to xDrip+ only once each 5 min, along side the 1 min raw.
So it's not each min even with that turned on.
I have it on.

**Comment by j-kaltes** [comment]
In old versions of xDrip the raw values are badly displayed. In a new version you can find all 1 minutes values between the 5 minutes values created by xDrip. I have xDrip version e13a76e-2022.08.03

**Comment by malloy139** [comment]
> > 
> 
> Jiggluco sends all the raw data to xDrip+ only once each 5 min, along side the 1 min raw. So it's not each min even with that turned on. I have it on.

This is not what I observe. You can see a new white dot every minute in xdrip.

**Comment by kaz45242** [comment]
Hi, just want to express my appreciation from US. I received Fl3 as sample from my Dr. I started using lb3 with juggluco and xdrip.  It is working correctly for two days for now. I experienced two incidents in relating to Bluetooth disconnection. One for phone battery power lost (4hrs)and other for swimming class(45min). Both cases, I could reconnect bluetooth with retouching phone to sensor. But I could not import those data to xdrip even though juggluco showing one of those data(for 45min). Is this legitimate condition?  I'm wondering whether I can connect watch and sensor directly?
But anyway I appreciate for those developers very much. 
 For your information I'm using following devises: Samson S22 phone and Garmin vivoactive3 watch.

**Comment by twinko** [comment]
@kaz45242 
Afaik there is no possibility to reimport missed data from juggluco to xdrip. This question was asked multiple times in the original juggluco thread (its in german), even though juggluco is capable of reading data from the sensor of the past 8h.

You can connect a smartwatch to the sensor directly if it has a wearOS, read here: http://jkaltes.byethost16.com/JugglucoWearOS/mgdL/index.html

If you are more into mibands or amazfit watches, there is no direct connection but its working flawless with watchdrip+ (it takes the data from. Xdrip) https://bigdigital.home.blog/

**Comment by henrycoding** [comment]
Any further information about Libre 3? Could we implement the sensor according to juggluco ? It seems that Juggluco is open source

**Comment by enpa23** [comment]
> Any further information about Libre 3? Could we implement the sensor according to juggluco ? It seems that Juggluco is open source

Juggluco code is open source but it needs proprietary libs in order to be built and run, so I don't think that a L3 direct connection could be ever implemented in xDrip.
Using juggluco as data source for xDrip is fine, the app is stable, frequently updated, and very light on battery and resources.
The only big missing feature is calibration because you have to set patched app source in order to collect data ( adding full calibration to that source would be very appreciated ).

Given that ( according to what the juggluco's devleper reports ) with L3 you don't have access to raw data, another source like a possible OOP3 or the Diabox app when it will enable its support would make no difference with the current solution.

**Comment by twinko** [comment]
@henrycoding this wont happen any time. Read explanation here: https://github.com/NightscoutFoundation/xDrip/issues/1474#issuecomment-1041327847

@enpa23 sorry but your information is only partially right :)

> The only big missing feature is calibration because you have to set patched app source in order to collect data ( adding full calibration to that source would be very appreciated ).

Please do not use the horrible patched libre app. Just scan the sensor with juggluco when you want to start a new one, thats it. 
I already updated the german version of the instructions here, but didnt had the time yet to translate it. https://github.com/maheini/FreeStyle-Libre-3-patch/blob/main/Juggluco-solution/instructions/de/instructions.md
Regarding calibration: Please read the part jka wrote at the bottom of the page about calibration: https://www.juggluco.nl/Juggluco/mgdL/index.html i dont understand why people want to calibrate a libre sensor. This is not a dexcom, calibration is not needed. And yes the data never matches if you prick your finger and compare the results, because the libre data is alway  15-20min older than the real blood glucose level. Thats how it is for all CGMs. Please google the reason for that by yourself :) 

> Given that ( according to what the juggluco's devleper reports ) with L3 you don't have access to raw data, another source like a possible OOP3 or the Diabox app when it will enable its support would make no difference with the current solution.

you are absolutely right, but to specifiy you can get a data point every minute

**Comment by enpa23** [comment]
@twinko 
I do use Juggluco, not the patched app. xDrip has to be set to patched app data source in order to receive data from it.

In my experience both L3 ( less ) and L2 ( way more ) often come bad calibrated, while still inside the error range that can entitle to a substitution from the vendor.
I easily have a 60 mg/dl difference with fingerprick on the 180-200 range, and so on, even when bg is stable for 15 minutes.
With a good calibration ( I have to enable experimental secondary plugins in xDrip, add few points at the beginning with stable bg - some in the ~100, some in the 180-200 ), add a new value after two days or so to check if sensor has stabilized to a different offset, and then in case check every few days ) I get almost identical values from xDrip and fingerprick test for the whole sensor life.
Switching from a so-so calibrated sensor to a close to perfect one has a great added value to me.

**Comment by klaaslusterbroodvan** [comment]
> @twinko I do use Juggluco, not the patched app. xDrip has to be set to patched app data source in order to receive data from it.
> 
> In my experience both L3 ( less ) and L2 ( way more ) often come bad calibrated, while still inside the error range that can entitle to a substitution from the vendor. I easily have a 60 mg/dl difference with fingerprick on the 180-200 range, and so on, even when bg is stable for 15 minutes. With a good calibration ( I have to enable experimental secondary plugins in xDrip, add few points at the beginning with stable bg - some in the ~100, some in the 180-200 ),

I agree, it makes sense to calibrate. Some of the L3 sensors are always given too high values. For those saying that differences are due to a lag between interstitial and blood glucose, there is one recent academic study by Alva and colleagues (2023) that shows it quite short (on average below 5 minutes). I personally found the calibration as is possible with Dexcom makes life easier as you can really rely. Good to hear that Twinko found the calibration working well, I haven't tried it myself with L3.
